Mathlib/Data/Finset/Update.lean

```diff
@@ -0,0 +1,61 @@
+/-
+Copyright (c) 2023 Floris van Doorn. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Floris van Doorn
+-/
+import Mathlib.Data.Finset.Basic
+
+/-!
+# Update a function on a set of values
+
+This file defines `Function.updateFinset`, the operation that updates a function on a
+(finite) set of values.
+-/
+variable {ι : Sort _} {π : ι → Sort _} {x : ∀ i, π i} [DecidableEq ι]
+
+namespace Function
+
+/-- `updateFinset x s y` is the vector `x` with the coordinates in `s` changed to the values of `y`.
+
+We should generalize `updateFinset` to `SetLike` instead of restricting to `Finset`.
+However, at the moment `Finset` is not `SetLike`, so we cannot do this yet.
+-/
+def updateFinset (x : ∀ i, π i) (s : Finset ι) (y : ∀ i : ↥s, π i) (i : ι) : π i :=
+  if hi : i ∈ s then y ⟨i, hi⟩ else x i
```

Comment:
It's not clear to me what the benefit of defining this on `Finset` is:
``suggestion
/-- `updateSet x s y` is the vector `x` with the coordinates in `s` changed to the values of `y`.

This is `Set.piecewise` where the left argument `x` is dependently-typed
-/
def updateSet (x : ∀ i, π i) (s : Set ι) [DecidablePred (· ∈ s)]  (y : ∀ i : ↥s, π i) (i : ι) : π i :=
  if hi : i ∈ s then y ⟨i, hi⟩ else x i
```

Reply:
I definitely want to apply this to `Finset`. I need to integrate over a `Pi`-type over this Finset at some point.

Your suggestion means that I need to coerce the `Finset` to `Set`. It also means that all lemmas I want are inconveniently stated in terms of `Set`-operations instead of `Finset`-operations. This makes it less convenient to work with.

The proper way to do this is to define this for all `SetLike` objects, but this is impossible at the moment, as the docstring describes.

Reply:
> So I need to add a bunch of rewrites commuting coercions with operations.

Shouldn't `norm_cast` handle this?

> The proper way to do this is to define this for all `SetLike` objects

I'd argue this is much worse, because now you need separate lemmas about `updateSet _ (s ∩ t)` for `Set` and `Finset`, and another seven or so for `updateSet _ (s ⊓ t)` for `Submonoid`, `Submodule`, .... Then you have to repeat this for `bot` and `iInf` and `sInf`.

If you do it for `set` you only need to write thes once.

Reply:
I gave this an honest try to use `updateSet` in our work. It is in https://github.com/leanprover-community/mathlib4/commit/d0aa7ef5d825085f84717d5eb8008c3910b90d3a but it a lot less nice.

In the theorem `marginal_union` I do the following calculation (`s` and `t` are Finsets):
```
(∫⋯∫_s ∪ t, f ∂μ) x = 
∫⁻ (y : (i : ↥(s ∪ t)) → π i), f (updateFinset x (s ∪ t) y) ∂.pi fun i' : ↥(s ∪ t) ↦ μ i'
```
This is true by definition when using `updateFinset`.

When using `updateSet`, I have to be very careful to use `Finset`-unions. If I use `Set`-unions, then there are multiple places where the expression becomes type-incorrect.

The rest of the calculation mostly goes through the same, if I'm very careful by explicitly stating where the coercions should be added. Some lemmas I still need to formulate for `Finset` specifically, and sometimes I can use `Set` analogues.

I strongly prefer to keep `updateFinset`. If you don't like this specificity, I'd be happy to call it `marginalAux` or something (that was actually the original name, until I thought that this operation might be more generally useful.